docs/agent_hocon_reference.md

```diff
@@ -0,0 +1,497 @@
+
+#### External Agents
+
+This is not a hocon file key, but more a description of a concept that relates to listings of tools.
+
+It is possible for any agent to reference another agent on the same server by adding a forward-slash
+in front of the served agent's name.  This is typically the stem of an agent network hocon file in
+a deployment's registries directory.
+
+Example: "/website_search" or "/math_guy"
+
+This allows common agent network definitions to be used as functions for other local networks.
+
+Furthermore, it is also possible to reference agents on other neuro-san _servers_ by using a URL as a tool reference.
+
+Example: "http://localhost:8080/math_guy"
+
+This enables entire ecosystems of agent webs.
+
+### llm_config
+
+It is possible for any LLM-enabled agent description to also have its own [llm_config](#llm-config)
+dictionary.  This allows for agents to use the right agent for the job.
+Some considerations might include:
+
+* Use of lower-cost LLMs for lighter (perhaps non-tool-using) jobs
+* Use of specially trained LLMs when subject matter expertise is required.
+* Use of securely sequestered LLMs when sensitive information is appropos to a single agent's chat stream.
+
+### class
+
+Optional string specifying a Python class which implements the
+[CodedTool](https://github.com/leaf-ai/neuro-san/blob/main/neuro_san/interfaces/coded_tool.py)
+interface.
+
+Implementations must be found in the directory where the class can be resolved by looking
+under the AGENT_TOOL_PATH environment variable setting as part of the PYTHONPATH.
+By default neuro-san deployments assume that PYTHONPATH is set to contain the
+top-level of your project's repo and that AGENT_TOOL_PATH is set to "<top-level>/coded_tools".
+In that directory each agent has its own folder and the value of the class is resolved
+from there.
+
+For example:
+If the agent is called "math_guy" and the class is valued as "calculator.Calculator",
+The python file math_guy/calculator.py under AGENT_TOOL_PATH is expected to have
+a class called Calculator which implements the CodedTool interface.
+
+
+Implementations of the CodedTool interface must have implementations which:
+
+* have a no-args constructor
+* implement either the preferred async_invoke() or the discouraged synchronous invoke() method.
+
+Agents representing CodedTools have the arguments described their [function parameters](#parameters)
+populated by calling LLMs and passed in via the args dictionary of their async/invoke() method
+when they are invoked.  They are also passed the sly_data dictionary which contains
+private information not accessable to the chat stream.
+
+Note that the CodedTool also has a synchronous invoke() method, but we discourage its use,
+as neuro-san is expected to run in an asynchronous multi-threaded environment.
+Using synchronous I/O calls within CodedTool implementations will result in loss of per-request
+agent parallelism and performance problems at scale.
```

Comment:
@Noravee : I don't have anything in here on base_tool stuff.
Seems like it's changing a lot and didn't want to expound about what I don't really know.
Will leave those docs to you.

FWIW: Leave toolbox hocon documentation to its own markdown file.
My intent is to have 1 md doc per hocon type.